Bump version to 0.0.1-alpha.12 (#700)
astral-sh/ty: CHANGELOG.md

# Changelog

## 0.0.1-alpha.12

### Bug fixes

- Avoid duplicate diagnostic when reporting errors in unpacked assignments ([#18897](https://github.com/astral-sh/ruff/pull/18897))
- Fix panics when "pulling types" for `ClassVar` or `Final` parameterized with >1 argument ([#18824](https://github.com/astral-sh/ruff/pull/18824)). These could cause issues when hovering over symbols in an IDE.

### Improved modeling of Python runtime semantics

- Add support for `@staticmethod`s ([#18809](https://github.com/astral-sh/ruff/pull/18809))

> - Assignments in the body of a staticmethod are never instance attributes ([#18587](https://github.com/astral-sh/ruff/pull/18587))

Reviewer: This PR also adds the ability to understand pure class attributes that are assigned in `@classmethod`s. This is probably the more interesting feature, the staticmethod thing is just a bugfix (that probably didn't affect anyone). So maybe
```suggestion
- Discover implicit class attribute assignments in `@classmethod`s ([#18587](https://github.com/astral-sh/ruff/pull/18587))
```

Author: thanks -- I think it is good to mention the staticmethod change, but I agree the classmethod change is more important! I applied a variant of this in https://github.com/astral-sh/ty/pull/700/commits/12a4194260759e4cb7fab128d400ea0ea94269ab

- Report when a dataclass contains more than one `KW_ONLY` field ([#18731](https://github.com/astral-sh/ruff/pull/18731))

### Type narrowing improvements

- Ty will now perform `isinstance()` and `issubclass()` narrowing when the second argument is a union type, intersection type or `TypeVar` type ([#18900](https://github.com/astral-sh/ruff/pull/18900))
- Ty now narrows types in comprehensions and generator expressions ([#18934](https://github.com/astral-sh/ruff/pull/18934))
- Understand two `NominalInstanceType`s as disjoint types if attempting to use multiple inheritance with their underlying classes would result in an instance memory layout conflict ([#18864](https://github.com/astral-sh/ruff/pull/18864))

### Other typing semantics features

- Support "mixed" tuples such as `tuple[int, *tuple[str, ...]]` ([#18600](https://github.com/astral-sh/ruff/pull/18600), [#18901](https://github.com/astral-sh/ruff/pull/18901))

> - Map subscript expression type inference over union types ([#18846](https://github.com/astral-sh/ruff/pull/18846))

Reviewer: Maybe
```suggestion
- Support type inference for subscript expressions on union types ([#18846](https://github.com/astral-sh/ruff/pull/18846))
```

> - Introduce a new subtyping framework such that gradual types can sometimes participate in subtyping ([#18799](https://github.com/astral-sh/ruff/pull/18799))

Reviewer: Maybe
```suggestion
- Introduce a new subtyping framework in which gradual types do participate, allowing for more advanced union type simplification ([#18799](https://github.com/astral-sh/ruff/pull/18799))
```
